# small-molecule-immune-training
Code implemented to analyze bulk ATAC-sequencing of BMDMs treated with novel small molecules from Knight et al. PNAS. (2024)

Visualizations of this analysis can be found in Figure 4, Supplementary Table 3, and Supplementary Figure 4.

Raw sequencing data is deposited in Geo. 


# 1 - Initial Installations
All ATAC-seq analysis was performed on the high performance research computing cluster Midway3 at the University of Chicago. To preprocess the raw sequencing, we began with command line installations found in 1-installations.sh. 

For subsequent analysis, we primarily applied the R programming language using version 4.1.0. Occasional implementation of python scripts or pip based installations used python-anaconda-2022.05-el8-x86_64. 

# 2 - Preprocessing


# 3 - Differential Accessibility Modeling

## limma

## mashr


# 4 - Gene Set Enrichment Analysis


# 5 - Differential Regulation of Transcription Factors

## Motif Enrichment

## Footprinting


# Scripts used to generate each figure panel

## Figure Four

### B

### C

### D

### E

## Supplementary Table One

## Supplementary Table Two

## Supplementary Table Three

## Supplementary Figure Four

### A

### B

### C

### D

### E

### F
